### PR TITLE
Feature/images

### PR DESCRIPTION
**Pull request by vanpelt**
@shawnlewis @adrianbg Here's the big refactor of the torch / history if you care to take a gander.  We should be able to start experimenting from this branch.

### CODE REVIEW

File: wandb/board/ui/src/components/PanelLinePlot.js [hunk 1]
```diff
@@ -106,7 +109,7 @@ class LinePlotPanel extends React.Component {
       return <p>This plot type is not supported on this page.</p>;
     }
     let data = this.props.data.history;
-    let keys = this.props.data.historyKeys;
+    let keys = this.props.data.historyKeys.filter(k => k !== 'examples');
```
Comment (shawnlewis) [inline wandb/board/ui/src/components/PanelLinePlot.js]: Is examples a special key?

Comment (vanpelt) [reply wandb/board/ui/src/components/PanelLinePlot.js]: Ahhh, it was for a brief moment in time.  Will fix today.

File: wandb/history.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        self._lock = Lock()
+        self._torch = None
+        try:
+            with open(self.fname) as f:
+                for line in f:
+                    self._index(json.loads(line))
+        except IOError:
+            pass
+
+        self._file = open(self.fname, 'w')
```
Comment (shawnlewis) [inline wandb/history.py]: Doesn't this truncate the file each time a stream is opened? Maybe 'a'?

File: wandb/history.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        """
+        if self.stream_name != "default":
+            raise ValueError("Nested streams aren't supported")
+        if self._streams.get(name) == None:
+            self._streams[name] = History(self.fname, out_dir=self.out_dir,
+                                          add_callback=self._add_callback, stream_name=name)
+        return self._streams[name]
+
+    def column(self, key):
+        """Fetches a key from all rows that have it. Skips those that don't.
```
Comment (shawnlewis) [inline wandb/history.py]: "Iterator over a giving column, skipping rows that don't have key"

File: wandb/history.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        # TODO: store a downsampled representation in memory
+        self.rows.append(row)
+        self._keys.update(row.keys())
+        self._steps += 1
+
+    def _transform(self):
+        """Transforms media classes into the proper format before writing"""
+        for key, val in six.iteritems(self.row):
+            if type(val) in (list, tuple):
+                if len(val) > 0 and type(val[0]) == media.Image:
```
Comment (shawnlewis) [inline wandb/history.py]: Maybe a stronger check?

```python
is_image = [type(v) == media.Image for v in val]
if all(is_image):
  # transform
elif any(is_image):
  raise SomeError('Lists may not contain mixed types')
```

or something like that.

File: wandb/history.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                    self.row[key] = media.Image.transform(val, self.out_dir,
+                                                          "{}_{}.jpg".format(key, self.row["_step"]))
+
+    def _write(self):
+        if self.row:
+            self._lock.acquire()
+            try:
+                self.row['_runtime'] = round(time.time() - self._start_time, 2)
```
Comment (shawnlewis) [inline wandb/history.py]: Should we save wall-clock time while we're at it? Would make it easy to make charts that work across runs.

Also, I know we talked about the rounding before but now I'm feeling like, why drop information here. We can round in the UI.

File: wandb/media.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
+import os
+from wandb import util
+
+
+class Media(object):
+    @staticmethod
+    def from_type(type):
```
Comment (shawnlewis) [inline wandb/media.py]: What's from_type?

If it's what I'm guessing it's probably better as a classmethod, which gives you the class as an argument (with a static method if you're generating instances of your class, you have to hardcode the class name within the class itself, which doesn't play as well with inheritance).

File: wandb/media.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+class Image(object):
+    def __init__(self, data, mode=None, caption=None):
+        """
+        Accepts numpy array of image data, or a PIL image. The class attempts to infer
+        the data format and converts it.
+        """
+        try:
+            from PIL import Image as PILImage
+        except ImportError:
+            raise ValueError(
```
Comment (shawnlewis) [inline wandb/media.py]: Cool I like this behavior of only requiring certain dependencies at runtime, although that raises the question. Do we put it in requirements.txt?

I think when you pip install a package you can pass arguments. So we could maybe have like pip install wandb --full or pip install wandb --barebones. Not sure though.

File: wandb/media.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        """
+        Guess what type of image the np.array is representing 
+        """
+        # TODO: do we want to support dimensions being at the beginning of the array?
+        if data.shape[-1] == 1:
+            return "L"
+        elif data.shape[-1] == 3:
+            return "RGB"
+        elif data.shape[-1] == 4:
+            return "RGBA"
```
Comment (shawnlewis) [inline wandb/media.py]: Should we raise in the else here? Otherwise we return None

File: wandb/media.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+    def to_uint8(self, data):
+        """
+        Converts floating point image on the range [0,1] and integer images
+        on the range [0,255] to uint8, clipping if necessary.
+        """
+        import numpy as np
```
Comment (shawnlewis) [inline wandb/media.py]: Maybe need the try/except around this too>?

File: wandb/history.py (hunk 4 above)
Comment (adrianbg) [reply wandb/history.py]: Better to use `instanceof(v, media.Image)` to work with inheritance

File: wandb/media.py (hunk 7 above)
Comment (vanpelt) [reply wandb/media.py]: Yeah, I've seen pip install urllib3[secure] maybe we could do something like that but for PyTorch, keras, TensorFlow, etc.

**Pull request merged**